Implement player events list scrolling behavior

alexkim205 opened the pull request:
## Changes

Builds off of this PR #6971 as part of the new recording playback revamp.

This PR more tightly couples the event list with the playback controller/frame and implements autoscrolling of the events list. 

This is now possible:

//V1: https://user-images.githubusercontent.com/13460330/141377760-77a7f3ae-d66b-49ff-9e1a-400331f644c7.mov
//V2:

https://user-images.githubusercontent.com/13460330/142083795-581296b7-02b1-4ba8-872e-77239be15218.mov

### Some areas that can be optimized:

- [x] ~~Autoscrolling faster than we can render~~ Workaround this with new "jump to current" behavior

//V1: https://user-images.githubusercontent.com/13460330/141376287-cf289b57-61ad-4876-a7e2-814a1caa379c.mov

Sunk too much time into this one so would like some guidance here. The video above shows what happens when we try to auto scroll faster than the DOM can render rows. This happens because virtualized lists make the tradeoff of rendering only the rows in the viewframe of the container. When a user scrolls away from the visible part of the list, the virtualized list will determine the new list of events to render using the new scroll position. This gives us a huge performance optimization when we're dealing with long lists of events, but leaves us with the dilemma of ugly loading states when we're autoscrolling too fast. The current workaround (that doesn't always when the scroll jump is too big) is to use [overscanning](https://github.com/bvaughn/react-virtualized/blob/master/docs/List.md#prop-types) which renders a specific number of rows (set to 50 right now) above and below the ones that are visible.

- [ ] Scrollbar jumping when dealing with long recordings with many events

We're fetching events in chunks, and we render them in the events list lazily upon arrival. This also means that `react-virtualized` autosizer has to measure the rows heights all over again on every `loadEventsSuccess` action which can get pretty expensive. This also has weird side effects such as the scrollbar jumping as more chunks get loaded in while the player is playing, or scrollbar drag not working/being delayed.

- [ ] Row height sizing isn't always accurate (and making it accurate could be expensive)

I think this might be of a shortcoming of `react-virtualized` but row height sizes aren't always accurate. Because row contains dynamic content, we measure the size of each cell to calculate the parent container's total height. Resizing the window a bit might knock these into place, but `clearRowCache` is an expensive process if we're dealing with a lot of rows.

## How did you test this code?

Add logic tests for any new functionality in `eventListLogic`

alexkim205 committed:
remove full screen and implement event list

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/player-events-recordings

alexkim205 committed:
implement event list

alexkim205 committed:
hide events list behind FF

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/player-events-recordings

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/player-events-recordings

alexkim205 committed:
events list card styles + fuse query options + more fixes

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/player-events-recordings

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/player-events-recordings

alexkim205 committed:
revert event item titles and fix responsiveness on mobile

alexkim205 committed:
event autoscrolls with player'

clarkus reviewed (dismissed):
This is so cool! My one concern aligns with your high speed playback example. That might be triggering for users who react physically to flashing or flickering images (https://developer.mozilla.org/en-US/docs/Web/Accessibility/Seizure_disorders). I wonder if we might show some placeholder spinner when this is happening? At rapid playback speeds that event list is going to be pretty useless. @paolodamico any thoughts on what to do here? 

We might also consider pausing autoplay when the user has otherwise modified that list offset. Say you hover over that and scroll up to view a different event. We wouldn't want to autoplay the user out of that context even if the recording was still playing.

alexkim205 commented:
I can see the stop autoscroll on hover over list behavior being super intuitive to use. 

As for the rapid playback speeds causing the events list to flicker, maybe we can disable the autoscroll feature for speeds greater than 4x? Something that could also help with list jittering would be to disable autoscrolling until all the events and snapshots are loaded. However, I'm not entirely sure what the best way to expose this to the user is.

clarkus commented:
Another idea might be paging the event list based on the playback speed. For example, only moving through in sets instead of rapidly for each item. That would mean less redraws, less movement, but still a representative playback experience.

alexkim205 committed:
feedback

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/player-events-recordings

alexkim205 committed:
Merge branch 'feat/player-events-recordings' into feat/couple-events-list-to-player

alexkim205 committed:
remove quotes to fix tokenization

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/player-events-recordings

alexkim205 committed:
Merge branch 'feat/player-events-recordings' into feat/couple-events-list-to-player

paolodamico commented:
My perspective:

1. I think intuitively if you're hovering over the list, the list should stop scrolling automatically but playback should continue, you may be looking for an event but still want to keep watching the recording.
2. For large playback speeds, I agree with Chris on moving by sections, and maybe we highlight a section instead of a specific event?

clarkus commented:
Another idea from this morning - a jump-to action that lets you catch up to the current events. Instead of autoscrolling, we auto-highlight and make it easy to jump to that context.

alexkim205 commented:
Thanks everyone for the feedback. I'll briefly summarize some of the points brought up here and offline.

- [ ] Remove autoscrolling
- [ ] Implement catchup mechanism to scroll to current event. Show action button whenever user scrolls current event out of frame.

<img width="326" alt="Screen Shot 2021-11-12 at 10 56 59 AM" src="https://user-images.githubusercontent.com/13460330/141520353-c3a14ace-8530-4a69-9f79-3c7943a0a9c0.png">

- [ ] Don't show events list until they're all loaded? Maybe same applies for seekbar events.
- [ ] Highlight sections of events with same timestamp, not specific events.
 
- ~~Stop autoscrolling on hover over list~~ - Irrelevant now that we're removing autoscroll
- ~~Paging events list~~ - Irrelevant now that we're removing autoscroll

alexkim205 committed:
fixing autoscroll

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/couple-events-list-to-player

alexkim205 committed:
fix timestampe css

alexkim205 committed:
remove autoscroll and animate highlight

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/couple-events-list-to-player

alexkim205 committed:
change start scripts for plugin-server

alexkim205 committed:
fix'

alexkim205 committed:
Merge branch 'fix/change-plugin-server-scripts-pycharm' into feat/couple-events-list-to-player

alexkim205 committed:
tweak anim and fix highlight bug

alexkim205 committed:
scroll to finder implement

alexkim205 committed:
refactor out row and cellrange renderers

alexkim205 commented:
Hi all, this PR is ready for review again. A few important things have changed since the first version, namely the way we keep the current events in frame. 

This revision also fixes a really annoying performance issue where entire DOM was rerendering when rrweb replays each snapshot. 
	- This turned out to be related to how often `react-virtualized` renders its cells. Fixed by overwriting rerender pattern with `useCallback`
	- This was also causing scrolling/jump to scrolling to be painfully slow.

alexkim205 committed:
add tests and optimize

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/couple-events-list-to-player

clarkus commented:
> When you try to jump to the current event while the player is playing through multiple events, you'll be taken to a stale scroll position that is no longer highlighted. This can be somewhat confusing for users who expect "jump to current" to mean "jump to current and stick with current" (kind of how Slack operates). Since we're not going to implement the "stick to current" functionality, I wonder if there's a better way to communicate this perceived lag. @clarkus

How about "jump to recent" or "jump to current time" - we're really putting them in the context of a timestamp - there could be multiple events. I don't think we need to be too targeted with our language - just something that communicates you can jump to the stuff you just saw play

alexkim205 committed:
ordering of event list item

alexkim205 committed:
change wording from current event to current time

alexkim205 committed:
eventually consistent row heights

clarkus commented:
> Highlight box height isn't always accurate. Same problem as "Row height sizing isn't always accurate (and making it accurate could be expensive)" in PR description.

Auto-capture feels like the exception here. It needs extra description to be meaningful and there is a very clear primary attribute for describing the event. We can either make this case more defined, or find a way to describe other events in similarly descriptive ways. For example, current URL is a great identifier for a pageview event. I think focusing on auto-capture makes sense for now, but eventually we should be able to get to a more predictable pattern for these items.

alexkim205 commented:
> I think focusing on auto-capture makes sense for now, but eventually we should be able to get to a more predictable pattern for these items.

From the implementation side, this would be amazing and so much quicker to render. Knowing the row heights for each event type beforehand (i.e., Autocapture -> 3 lines, pageevents -> 2 lines, etc.) would mean `react-virtualized` can precalculate the total height without `<CellMeasurer/>` bloat. I'll refactor some of this so that we can add/remove/change any events we see fit into the an eventsToRowHeights map

clarkus commented:
> From the implementation side, this would be amazing and so much quicker to render. Knowing the row heights for each event type beforehand (i.e., Autocapture -> 3 lines, pageevents -> 2 lines, etc.) would mean `react-virtualized` can precalculate the total height without `<CellMeasurer/>` bloat. I'll refactor some of this so that we can add/remove/change any events we see fit into the an eventsToRowHeights map

Brought up some examples in figma to help work through this. https://www.figma.com/file/gQBj9YnNgD8YW4nBwCVLZf?node-id=5186:36628#125681661

rcmarron reviewed (dismissed):
Looking great - I tried to find bugs but it was hard! A couple small things:
* We might want a bit of margin at the top of the list. The blue highlight goes right into the search box.
![image](https://user-images.githubusercontent.com/4813045/142462019-9e2d4fac-1188-4fee-8815-a0b1c9b60a7a.png)
* The event loading indicator never seems to appear.
* Do we want to handle the "no events" case. So showing some message like "No events fired in this recording".

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/couple-events-list-to-player

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/couple-events-list-to-player

alexkim205 commented on frontend/src/scenes/session-recordings/player/PlayerEvents.tsx:
```diff
@@ -1,66 +1,144 @@
+                        </Col>
+                        <Col
+                            className={clsx('event-item-content', {
+                                rendering: !renderAllRows && !isRowIndexRendered(index),
```
good catch, looks like i overthought this one.

alexkim205 committed:
refactor util

alexkim205 commented:
Thanks for the reviews. I'm going to offload the row heights work into a smaller PR to get this one in quickly. @rcmarron addressed your feedback! Now events shouldn't show up in the list until all the events are loaded. A loading indicator takes its place and this also helps with any scroll jumping that we had before. Empty and loading states are also properly separated out.

alexkim205 committed:
fix empty state

rcmarron reviewed (approved):
LGTM!

One last little thing. It looks like the events list is staying in loading until the entire recording is loaded - not just the events.

alexkim205 commented:
> LGTM!
> 
> One last little thing. It looks like the events list is staying in loading until the entire recording is loaded - not just the events.

Hm I did that to prevent the empty data state from showing and flickering to the Loading indicator on opening a recording, but maybe the better solution is to show the loading only when eventsDataIsLoading && firstSnapshotChunkIsLoading.

alexkim205 committed:
loading
